web/app/js/components/Top.jsx

```diff
@@ -0,0 +1,378 @@
+  }
+
+  topEventKey = event => {
+    return [event.source.str, event.destination.str, event.http.requestInit.path].join("_");
+  }
+
+  initialTopResult(d, eventKey) {
+    return {
+      count: 1,
+      best: d.responseEnd.latency,
+      worst: d.responseEnd.latency,
+      last: d.responseEnd.latency,
+      success: !d.success ? 0 : 1,
+      failure: !d.success ? 1 : 0,
+      successRate: !d.success ? new Percentage(0, 1) : new Percentage(1, 1),
+      source: d.requestInit.source,
+      sourceLabels: d.requestInit.sourceMeta.labels,
+      destination: d.requestInit.destination,
+      destinationLabels: d.requestInit.destinationMeta.labels,
+      path: d.requestInit.http.requestInit.path,
+      key: eventKey,
+      lastUpdated: Date.now()
+    };
+  }
+
+  incrementTopResult(d, result) {
+    result.count++;
+    if (!d.success) {
+      result.failure++;
+    } else {
+      result.success++;
+    }
+    result.successRate = new Percentage(result.success, result.success + result.failure);
+
+    result.last = d.responseEnd.latency;
+    if (d.responseEnd.latency < result.best) {
+      result.best = d.responseEnd.latency;
+    }
+    if (d.responseEnd.latency > result.worst) {
+      result.worst = d.responseEnd.latency;
+    }
+
+    result.lastUpdated = Date.now();
+  }
+
+  indexTopResult = (d, topResults) => {
+    let eventKey = this.topEventKey(d.requestInit);
+    this.addSuccessRate(d);
+
+    if (!topResults[eventKey]) {
+      topResults[eventKey] = this.initialTopResult(d, eventKey);
+    } else {
+      this.incrementTopResult(d, topResults[eventKey]);
+    }
+
+    if (_.size(topResults) > this.state.maxRowsToStore) {
+      this.deleteOldestIndexedResult(topResults);
+    }
+
+    return topResults;
+  }
+
+  indexTapResult = data => {
+    // keep an index of tap results by id until the request is complete.
+    // when the request has completed, add it to the aggregated Top counts and
+    // discard the individual tap result
+    let resultIndex = this.state.tapResultsById;
+    let d = this.parseTapResult(data);
+
+    if (_.isNil(resultIndex[d.id])) {
+      // don't let tapResultsById grow unbounded
+      if (_.size(resultIndex) > this.state.maxRowsToStore) {
+        this.deleteOldestIndexedResult(resultIndex);
+      }
+
+      resultIndex[d.id] = {};
+    }
+    resultIndex[d.id][d.eventType] = d;
+
+    // assumption: requests of a given id all share the same high level metadata
+    resultIndex[d.id].base = d;
+    resultIndex[d.id].lastUpdated = Date.now();
+
+    let topIndex = this.state.topEventIndex;
+    if (d.completed) {
+      // only add results into top if the request has completed
+      // we can also now delete this result from the Tap result index
+      topIndex = this.indexTopResult(resultIndex[d.id], topIndex);
+      delete resultIndex[d.id];
+    }
+
+    this.setState({
+      tapResultsById: resultIndex,
+      topEventIndex: topIndex
+    });
+  }
+
+  addSuccessRate = d => {
```

Comment:
Super minor, but I'd call this method `addSuccessCount` or something similar. It's not actually computing the success rate -- it's adding the success count so that the success rate can be computed elsewhere.

Reply:
it's true!!!